core/teams/engine.go

```diff
@@ -0,0 +1,148 @@
+package teams
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"code.vegaprotocol.io/vega/core/types"
+	proto "code.vegaprotocol.io/vega/protos/vega"
+)
+
+var ErrReferrerCannotJoinAnotherTeam = errors.New("a referrer cannot join another team")
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/mocks.go -package mocks code.vegaprotocol.io/vega/core/teams EpochEngine
+
+type EpochEngine interface {
+	NotifyOnEpoch(func(context.Context, types.Epoch), func(context.Context, types.Epoch))
+}
+
+type Engine struct {
+	// teams tracks all teams by team ID.
+	teams map[types.TeamID]*types.Team
+
+	// allTeamMembers tracks all the parties that belongs to a team, referrers and
+	// referees, by their current team ID.
+	allTeamMembers map[types.PartyID]types.TeamID
+
+	// membersToMove tracks all the parties that switch teams. The switch only
+	// happens by the end of the epoch.
+	membersToMove map[types.PartyID]teamMove
+}
+
+func (e *Engine) CreateTeam(referrer types.PartyID, name string, teamURL string, avatarURL string, enableRewards bool) (types.TeamID, error) {
+	newTeamID := e.newUniqueTeamID()
```

Comment:
This is fine for now, but I would think that later on this will be passed as a parameter and be a deterministic ID extract from the transaction as for every ID in the system.

Reply:
alright